csrc/multidevice/communication.cpp

```diff
@@ -97,54 +99,101 @@ T getInitialValue(c10d::ReduceOp::RedOpType op) {
 }
 
 bool hasRoot(CommunicationType type) {
-  return type == CommunicationType::Gather ||
-      type == CommunicationType::Scatter || type == CommunicationType::Reduce ||
-      type == CommunicationType::Broadcast ||
-      type == CommunicationType::SendRecv;
+  switch (type) {
+    case CommunicationType::Gather:
+    case CommunicationType::Scatter:
+    case CommunicationType::Reduce:
+    case CommunicationType::Broadcast:
+    case CommunicationType::SendRecv:
+      return true;
+    case CommunicationType::Allgather:
+    case CommunicationType::Allreduce:
+    case CommunicationType::ReduceScatter:
+      return false;
+    default:
+      NVF_ERROR(false, "unrecognized CommunicationType: ", type);
+  }
 }
 
 bool isReduction(CommunicationType type) {
-  return type == CommunicationType::Reduce ||
-      type == CommunicationType::Allreduce ||
-      type == CommunicationType::ReduceScatter;
+  switch (type) {
+    case CommunicationType::Reduce:
+    case CommunicationType::Allreduce:
+    case CommunicationType::ReduceScatter:
+      return true;
+    case CommunicationType::Gather:
+    case CommunicationType::Allgather:
+    case CommunicationType::Scatter:
+    case CommunicationType::Broadcast:
+    case CommunicationType::SendRecv:
+      return false;
+    default:
+      NVF_ERROR(false, "unrecognized CommunicationType: ", type);
+  }
 }
 
 } // namespace
 
 Communication::Communication(
     IrBuilderPasskey passkey,
     CommunicationType type,
-    DeviceMesh mesh,
+    TensorView* out,
+    TensorView* in,
     Team team,
     DeviceIdxType root,
     RedOpType red_op,
-    int64_t scattered_axis,
-    TensorView* input_tv,
-    TensorView* output_tv)
+    int64_t scattered_axis)
     : Expr(passkey) {
-  NVF_ERROR(mesh.size() > 0, "The mesh size must be greater than 0.");
   NVF_ERROR(
-      hasRoot(type) == (root >= 0),
-      "Root ",
-      root,
-      " is not expected by CommunicationType ",
-      type);
-  NVF_ERROR(isReduction(type) == (red_op != RedOpType::UNUSED))
+      in->getDeviceMesh().size() > 0,
+      "The input mesh size must be greater than 0.");
   NVF_ERROR(
-      (type == CommunicationType::ReduceScatter) == (scattered_axis >= 0));
+      out->getDeviceMesh().size() > 0,
+      "The output mesh size must be greater than 0.");
+
+  addInput(in);
+  addOutput(out);
+  addDataAttribute(type);
```

Comment:
why not calling the other constructor here instead of duplicating?

Reply:
Good point. I tried constructor delegation but couldn't figure out a good way to call `validate` and also check the mesh -- I want both constructors to call the same `validate` but check meshes differently. I'll leave it as is for now, because we'll eventually (or soon?) remove the mesh.